src/hooks/useFetchWeather.js

```diff
@@ -0,0 +1,109 @@
+import { useState, useEffect } from "react";
+import { mapCurrentWeatherData, mapForecastData } from "../lib/mappers";
+import { WEATHER_TYPES, API, UNITS } from "../constants";
+import { API_KEY } from "../config";
+
+const useFetchWeather = (initialLocation) => {
+  const [currentWeatherData, setCurrentWeatherData] = useState({});
+  const [forecastWeatherData, setForecastWeatherData] = useState([]);
+  const [error, setError] = useState(null);
+  const [unit, setUnit] = useState(UNITS.CELSIUS);
+  const [isLoading, setIsLoading] = useState(true);
+
+  const generateUrl = (location, type, apiKey) => {
+    const baseEndpoint =
+      type === WEATHER_TYPES.CURRENT ? API.CURRENT_URL : API.FORECAST_URL;
+    let url = "";
+
+    if (typeof location === "string") {
+      url = `${baseEndpoint}?q=${location}&appid=${apiKey}&units=metric`;
+    } else if (typeof location === "object" && location.lat && location.lon) {
+      url = `${baseEndpoint}?lat=${location.lat}&lon=${location.lon}&appid=${apiKey}&units=metric`;
+    } else {
+      throw new Error("Invalid location format");
+    }
+
+    return url;
+  };
```

Comment:
Tip 1: Remember what I said about let and reassigning variables? Always when you see that, try to avoid, because it's a code smell.

Tip 2: try to avoid very nested if/else if/else.

Tip 3: you don't need to pass `apiKey` as parameter, because it will be always the same constant.

Tip 4: sometimes if it's one or another, you just need to do the first comparison, so I created a helper variable to create the locationQuery and this simplified even more my url, being easier to read.

```suggestion
  const buildWeatherUrl = (location, type) => {
    const baseEndpoint =
      type === WEATHER_TYPES.CURRENT ? API.CURRENT_URL : API.FORECAST_URL;
    const locationQuery =
      typeof location === "string"
        ? `q=${location}`
        : `lat=${location.lat}&lon=${location.lon}`;
    return `${baseEndpoint}?${locationQuery}&appid=${API_KEY}&units=metric`;
  };
```